File: kinova_python/kinova_python/joint_mode.py
import rclpy
from rclpy.node import Node
from controller_manager_msgs.srv import SwitchController

class ControllerSwitchNode(Node):
    def __init__(self):
        super().__init__('controller_switch_node')
        self.client = self.create_client(SwitchController, '/controller_manager/switch_controller')
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for service /controller_manager/switch_controller...')

    def switch_controllers(self):
        request = SwitchController.Request()
        request.activate_controllers = ['gen3_lite_joint_trajectory_controller']
        request.deactivate_controllers = ['twist_controller']
        request.strictness = 1
        request.activate_asap = True

        # 컨트롤러 전환 요청
        future = self.client.call_async(request)
        rclpy.spin_until_future_complete(self, future)

        if future.result() is not None:
            self.get_logger().info('Controllers switched successfully')
        else:
            self.get_logger().error('Failed to switch controllers')

        self.get_logger().info('Shutting down node...')
        self.destroy_node()
        rclpy.shutdown()
        
def main(args=None):
    rclpy.init(args=args)
    controller_switch_node = ControllerSwitchNode()
    controller_switch_node.switch_controllers()
    rclpy.spin(controller_switch_node)

if __name__ == '__main__':
    main()

File: kinova_cpp/src/joint_mode.cpp
#include "rclcpp/rclcpp.hpp"
#include "controller_manager_msgs/srv/switch_controller.hpp"

class ControllerSwitchNode : public rclcpp::Node
{
public:
    ControllerSwitchNode()
        : Node("controller_switch_node")
    {
        // SwitchController 서비스 클라이언트 생성
        client_ = this->create_client<controller_manager_msgs::srv::SwitchController>("/controller_manager/switch_controller");

        // 서비스가 활성화될 때까지 기다리기
        while (!client_->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_INFO(this->get_logger(), "Waiting for service /controller_manager/switch_controller...");
        }
    }

    void switch_controllers()
    {
        // SwitchController 서비스 요청 생성
        auto request = std::make_shared<controller_manager_msgs::srv::SwitchController::Request>();
        request->activate_controllers = {"gen3_lite_joint_trajectory_controller"};
        request->deactivate_controllers = {"twist_controller"};
        request->strictness = 1;
        request->activate_asap = true;

        // 서비스 호출
        auto future = client_->async_send_request(request);

        // 서비스 결과 기다리기
        if (rclcpp::spin_until_future_complete(shared_from_this(), future) == rclcpp::FutureReturnCode::SUCCESS) {
            RCLCPP_INFO(this->get_logger(), "Controllers switched successfully");
        } else {
            RCLCPP_ERROR(this->get_logger(), "Failed to switch controllers");
        }

        // 노드 종료
        RCLCPP_INFO(this->get_logger(), "Shutting down node...");
        rclcpp::shutdown(); // shutdown을 여기서 호출
    }

private:
    rclcpp::Client<controller_manager_msgs::srv::SwitchController>::SharedPtr client_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto controller_switch_node = std::make_shared<ControllerSwitchNode>();
    controller_switch_node->switch_controllers();
    // spin 호출 제거
    return 0;
}
